### PR TITLE
[tests] add 5_7_01 network diagnostic test cases

### PR DESCRIPTION
**Pull request by wgtdkp**
This PR adds Network Diagnostic TLV parsing stuff and 5.7.01 test case. More Network Diagnostic test cases may come in following PRs.

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_7_01_CoapDiagCommands_A.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        self.simulator.go(5)
+        self.assertEqual(self.nodes[FED1].get_state(), 'child')
+
+        dut_rloc = self.nodes[DUT].get_ip6_address(config.ADDRESS_TYPE.RLOC)
+
+        # 2 - Leader sends DIAG_GET.req
+        tlv_types = [
+            TlvType.EXT_ADDRESS, TlvType.ADDRESS16, TlvType.MODE,
+            TlvType.CONNECTIVITY, TlvType.ROUTE64, TlvType.LEADER_DATA,
+            TlvType.NETWORK_DTA, TlvType.IPV6_ADDRESS_LIST,
```
Comment (bukepo) [inline tests/scripts/thread-cert/Cert_5_7_01_CoapDiagCommands_A.py]:
```suggestion
            TlvType.NETWORK_DATA, TlvType.IPV6_ADDRESS_LIST,
```

File: tests/scripts/thread-cert/network_diag.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    MAC_COUNTERS = 9
+    BATTERY_LEVEL = 14
+    SUPPLY_VOLTAGE = 15
+    CHILD_TABLE = 16
+    CHANNEL_PAGES = 17
+    TYPE_LIST = 18
+    MAX_CHILD_TIMEOUT = 19
+
+
+class Ipv6AddressList(object):
```
Comment (bukepo) [inline tests/scripts/thread-cert/network_diag.py]: no need for inherit from `object`?
```suggestion
class Ipv6AddressList:
```

File: tests/scripts/thread-cert/network_diag.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    def counters(self):
+        return self._counters
+
+    def __eq__(self, other):
+        common.expect_the_same_class(self, other)
+
+        return self.counters == other.counters
+
+    def __repr__(self):
+        return ("MacCounters(" +
```
Comment (bukepo) [inline tests/scripts/thread-cert/network_diag.py]: Unify using single quote?

File: tests/scripts/thread-cert/network_diag.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+#
+
+import struct
+
+import common
+import ipaddress
+import mle
+import network_layer
+
+from enum import IntEnum
```
Comment (bukepo) [inline tests/scripts/thread-cert/network_diag.py]: move this above as it's standard imports.

File: tests/scripts/thread-cert/network_diag.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    SUPPLY_VOLTAGE = 15
+    CHILD_TABLE = 16
+    CHANNEL_PAGES = 17
+    TYPE_LIST = 18
+    MAX_CHILD_TIMEOUT = 19
+
+
+class Ipv6AddressList:
+
+    def __init__(self, addresses: [ipaddress.IPv6Address]):
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/network_diag.py]:
```suggestion
    def __init__(self, addresses: typing.List[ipaddress.IPv6Address]):
```

File: tests/scripts/thread-cert/network_diag.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    def parse(self, data, message_info):
+        addresses = []
+        while data.tell() < message_info.length:
+            addresses.append(ipaddress.IPv6Address(data.read(16)))
+        return Ipv6AddressList(addresses)
+
+
+class MacCounters:
+
+    def __init__(self, counters: [int]):
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/network_diag.py]:
```suggestion
    def __init__(self, counters: typing.List[int]):
```
I would suggest to use `tuple` if counters are of fixed length.

File: tests/scripts/thread-cert/network_diag.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+                f'if_out_discards={self.if_out_discards})')
+
+
+class MacCountersFactory:
+
+    def parse(self, data, message_info):
+        return MacCounters(struct.unpack('>9I', data.read(4 * 9)))
+
+
+class BatteryLevel:
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/network_diag.py]: why not just use `int`?

File: tests/scripts/thread-cert/network_diag.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        return (self.timeout == other.timeout and
+                self.child_id == other.child_id and self.mode == other.mode)
+
+    def __repr__(self):
+        return f'ChildTableEntry(timeout={self.timeout}, child_id={self.child_id}, mode={self.mode})'
+
+
+class ChildTable:
+
+    def __init__(self, children: [ChildTableEntry]):
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/network_diag.py]: ditto

File: tests/scripts/thread-cert/network_diag.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+        self._children = children
+
+    @property
+    def children(self):
+        return self._children
+
+    def __eq__(self, other):
+        common.expect_the_same_class(self, other)
+
+        return self.children == other.children
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/network_diag.py]: Should same children info but different order be considered equal?  It is not currently.

File: tests/scripts/thread-cert/network_diag.py [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+
+class ChannelPagesFactory:
+
+    def parse(self, data, message_info):
+        return ChannelPages(data.getvalue())
+
+
+class TypeList:
+
+    def __init__(self, tlv_types: [int]):
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/network_diag.py]: ditto

File: tests/scripts/thread-cert/network_diag.py (hunk 7 above)
Comment (wgtdkp) [reply tests/scripts/thread-cert/network_diag.py]: Many other single field TLVs defines a class for it. So I'd like to keep the same approach.

File: tests/scripts/thread-cert/network_diag.py (hunk 9 above)
Comment (wgtdkp) [reply tests/scripts/thread-cert/network_diag.py]: Resolved by sorting children by ID.

File: tests/scripts/thread-cert/network_diag.py (hunk 6 above)
Comment (wgtdkp) [reply tests/scripts/thread-cert/network_diag.py]: list[int] is more easier for the parser and people are much more familiar with list :)

**Comment by bukepo** [comment]
ping~

**Comment by jwhui** [comment]
> ping~

@wgtdkp , build checks are failing - any plans to move this PR forward?

**Comment by wgtdkp** [comment]
@jwhui yes, will fix later today.

**Comment by jwhui** [comment]
> @jwhui yes, will fix later today.

Thanks!

**Comment by wgtdkp** [comment]
@jwhui This PR is ready for your review now~

File: tests/scripts/thread-cert/config.py [hunk 11]
```diff
@@ -14,7 +14,7 @@
 #     names of its contributors may be used to endorse or promote products
 #     derived from this software without specific prior written permission.
 #
-#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
```
Comment (jwhui) [inline tests/scripts/thread-cert/config.py:17]:
```suggestion
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
```

File: tests/scripts/thread-cert/network_diag.py [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
```
Comment (jwhui) [inline tests/scripts/thread-cert/network_diag.py:17]:
```suggestion
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
```

File: tests/scripts/thread-cert/Cert_5_7_01_CoapDiagCommands_A.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_7_01_CoapDiagCommands_A.py:17]:
```suggestion
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
```

**Pull request merged**